pkg/agent/podresapi/client.go

```diff
@@ -0,0 +1,167 @@
+// Copyright The NRI Plugins Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package podresapi
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strings"
+	"time"
+
+	logger "github.com/containers/nri-plugins/pkg/log"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+
+	api "k8s.io/kubelet/pkg/apis/podresources/v1"
+)
+
+// ClientOption is an option for the client.
+type ClientOption func(*Client)
+
+// Client is Pod Resources API client.
+type Client struct {
+	socketPath string
+	conn       *grpc.ClientConn
+	cli        api.PodResourcesListerClient
+	noGet      bool
+}
+
+const (
+	// these constants were obtained from NFD sources, cross-checked against
+	//   https://github.com/kubernetes/kubernetes/blob/release-1.31/test/e2e_node/util.go#L83
+	defaultSocketPath = "/var/lib/kubelet/pod-resources/kubelet.sock"
+	timeout           = 10 * time.Second
```

Comment:
shall we reduce timeout in future? it might be too long for our operations.

Reply:
This timeout is now only used for establishing a client connection. For actually listing all pod resources or querying a single one the caller provides `time.Duration` timeout (in the agent interface) or a `context.Context` which can then be a `context.WithTimeout` to use a timeout for the call.

Maybe that const should be/have been called a `connectionTimeout` ?

Reply:
Asynchronous List()s/Get()s we fire off with a 2 and 1 second timeouts respectively. However (these could also much longer), since the async. functions return a channel that can be used to wait for and eventually obtain the result, it is possible to programmatically override with a simple select with a channel receive and a time.After receive cases.